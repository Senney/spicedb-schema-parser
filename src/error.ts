class LexerError extends Error {
  constructor(line: number, index: number, message: string) {
    super(`${line}:${index}: ${message}`);
  }
}

class SyntaxError extends Error {
  constructor(line: number, index: number, message: string) {
    super(`${line}:${index}: ${message}`);
  }
}

export { LexerError, SyntaxError };
