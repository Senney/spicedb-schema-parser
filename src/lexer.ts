/**
 * Note: Caveats will be implemented later.
 */

import { LexerError } from './error';

export enum TokenType {
  IDENTIFIER,
  DEFINITION,
  RELATION,
  PERMISSION,

  ASSIGN, // =
  EQUAL, // ==
  UNION, // +
  INTERSECTION, // &
  EXCLUDE, // -
  OR, // |
  ARROW, // ->
  MEMBER, // #

  COLON, // :
  LEFT_BRACE, // {
  RIGHT_BRACE, // }
  LEFT_PAREN, // (
  RIGHT_PAREN, // )

  EOF,
}

const spiceDBKeywordLexemes: Record<string, TokenType> = {
  definition: TokenType.DEFINITION,
  relation: TokenType.RELATION,
  permission: TokenType.PERMISSION,
};

export class Token {
  constructor(
    public readonly token: TokenType,
    public readonly lexeme: string,
    public readonly literal: string | null,
    public readonly line: number,
    public readonly col: number,
  ) {}
}

const validIdentifierChars = /[a-z\_][a-z\_\/]*/;

class SpiceDBSchemaLexer {
  private start = 0;
  private current = 0;
  private line = 1;

  constructor(private readonly source: string) {}

  public scan(): Token {
    if (this.isAtEnd()) {
      return new Token(TokenType.EOF, '', null, this.line, this.current);
    }

    const c = this.advance();

    switch (c) {
      case '{':
        return this.emit(this.syntaxToken(TokenType.LEFT_BRACE));
      case '}':
        return this.emit(this.syntaxToken(TokenType.RIGHT_BRACE));
      case '(':
        return this.emit(this.syntaxToken(TokenType.LEFT_PAREN));
      case ')':
        return this.emit(this.syntaxToken(TokenType.RIGHT_PAREN));
      case ':':
        return this.emit(this.syntaxToken(TokenType.COLON));
      case '=':
        if (this.match('=')) {
          return this.emit(this.syntaxToken(TokenType.EQUAL));
        }

        return this.emit(this.syntaxToken(TokenType.ASSIGN));
      case '+':
        return this.emit(this.syntaxToken(TokenType.UNION));
      case '&':
        return this.emit(this.syntaxToken(TokenType.INTERSECTION));
      case '|':
        return this.emit(this.syntaxToken(TokenType.OR));
      case '-':
        if (this.match('>')) {
          return this.emit(this.syntaxToken(TokenType.ARROW));
        }

        return this.emit(this.syntaxToken(TokenType.EXCLUDE));
      case '#':
        return this.emit(this.syntaxToken(TokenType.MEMBER));
      case ' ':
      case '\r':
      case '\t':
        return this.dive();
      case '\n':
        this.line++;
        return this.dive();
      case '/':
        if (this.match('*')) {
          while (
            !(this.advance() === '*' && this.peek() === '/') &&
            !this.isAtEnd()
          ) {
            let a = this.peek();
          }
          this.advance();
          return this.dive();
        }

        if (this.match('/')) {
          while (this.advance() !== '\n' && !this.isAtEnd()) {}
          return this.dive();
        }
      default:
        if (validIdentifierChars.test(c)) {
          return this.emit(this.identifier());
        }
    }

    throw new LexerError(this.line, this.start, `Unrecognized token "${c}".`);
  }

  public scanAll(): Token[] {
    const tokens = [];
    for (
      let token = this.scan();
      token.token !== TokenType.EOF;
      token = this.scan()
    ) {
      tokens.push(token);
    }
    return tokens;
  }

  private isAtEnd(): boolean {
    return this.current >= this.source.length;
  }

  private advance(): string {
    return this.source.charAt(this.current++);
  }

  private peek(): string {
    return this.source.charAt(this.current);
  }

  private match(c: string): boolean {
    if (this.peek() === c) {
      this.advance();

      return true;
    }

    return false;
  }

  private identifier(): Token {
    while (validIdentifierChars.test(this.peek())) {
      this.advance();
    }

    const value = this.source.slice(this.start, this.current);
    const tokenType = spiceDBKeywordLexemes[value] ?? TokenType.IDENTIFIER;

    return new Token(tokenType, value, value, this.line, this.current);
  }

  private emit(token: Token): Token {
    this.start = this.current;

    return token;
  }

  private dive(): Token {
    this.start = this.current;

    return this.scan();
  }

  private syntaxToken(type: TokenType): Token {
    return new Token(
      type,
      this.source.slice(this.start, this.current),
      null,
      this.line,
      this.current,
    );
  }
}

export { SpiceDBSchemaLexer };
