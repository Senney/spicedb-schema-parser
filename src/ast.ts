import { Token } from './lexer';

class Statement {}
class Expression {}

class DefinitionStatement extends Statement {
  constructor(
    public resourceName: string,
    public block: BlockStatement,
  ) {
    super();
  }
}

class BlockStatement extends Statement {
  constructor(public statements: Statement[]) {
    super();
  }
}

class RelationStatement extends Statement {
  constructor(
    public relationName: string,
    public relationDefinition: RelationDefinitionExpression,
  ) {
    super();
  }
}

class PermissionStatement extends Statement {
  constructor(
    public permissionName: string,
    public permissionDefinition: PermissionDefinitionExpression,
  ) {
    super();
  }
}

class RelationDefinitionExpression extends Expression {
  constructor() {
    super();
  }
}

class PermissionDefinitionExpression extends Expression {
  constructor() {
    super();
  }
}

class BinaryExpression extends Expression {
  constructor(
    public lhs: Expression,
    public operator: Token,
    public rhs: Expression,
  ) {
    super();
  }
}

class GroupingExpression extends Expression {
  constructor(public expression: Expression) {
    super();
  }
}

export {
  BinaryExpression,
  BlockStatement,
  DefinitionStatement,
  Expression,
  GroupingExpression,
  PermissionDefinitionExpression,
  PermissionStatement,
  RelationDefinitionExpression,
  RelationStatement,
  Statement,
}
