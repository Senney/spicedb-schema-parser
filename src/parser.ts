import {
  BinaryExpression,
  BlockStatement,
  DefinitionStatement,
  Expression,
  IdentifierExpression,
  MembershipIdentifierExpression,
  PermissionDefinitionExpression,
  PermissionStatement,
  RelationDefinitionExpression,
  RelationStatement,
  Statement,
} from './ast';
import { SyntaxError } from './error';
import { Token, TokenType } from './lexer';

class SpiceDBSchemaParser {
  private current = 0;

  constructor(private tokens: Token[]) {}

  public parse(): Statement[] {
    this.current = 0;

    const statements: Statement[] = [];

    while (this.current < this.tokens.length) {
      statements.push(this.definitionStatement());
    }

    return statements;
  }

  private definitionStatement(): Statement {
    this.consume(TokenType.DEFINITION, 'definition');
    const resourceName = this.consume(TokenType.IDENTIFIER, 'identifier');

    return new DefinitionStatement(resourceName.lexeme, this.blockStatement());
  }

  private blockStatement(): BlockStatement {
    this.consume(TokenType.LEFT_BRACE, '{');

    const statements: Statement[] = [];

    while (!this.check(TokenType.RIGHT_BRACE)) {
      statements.push(this.relationStatement());
    }

    this.consume(TokenType.RIGHT_BRACE, '}');

    return new BlockStatement(statements);
  }

  private relationStatement(): Statement {
    if (!this.check(TokenType.RELATION)) {
      return this.permissionStatement();
    }

    this.consume(TokenType.RELATION, 'relation');
    const identifierToken = this.consume(TokenType.IDENTIFIER, 'identifier');
    this.consume(TokenType.COLON, ':');

    return new RelationStatement(
      identifierToken.lexeme,
      new RelationDefinitionExpression(this.relationDefinitionExpression()),
    );
  }

  private permissionStatement(): Statement {
    this.consume(TokenType.PERMISSION, 'permission');

    return new PermissionStatement(
      this.consume(TokenType.IDENTIFIER, 'identifier').lexeme,
      new PermissionDefinitionExpression(),
    );
  }

  private relationDefinitionExpression(): Expression {
    return this.relationOr();
  }

  private relationOr(): Expression {
    const lhs = this.relationIdentifier();

    if (this.match(TokenType.OR)) {
      const operator = this.previous();
      const rhs = this.relationDefinitionExpression();

      return new BinaryExpression(lhs, operator, rhs);
    }

    return lhs;
  }

  private relationIdentifier(): Expression {
    const symbol = this.consume(TokenType.IDENTIFIER, 'identifier');

    if (this.match(TokenType.MEMBER)) {
      const membership = this.consume(TokenType.IDENTIFIER, 'identifier');

      return new MembershipIdentifierExpression(symbol, membership);
    }

    return new IdentifierExpression(symbol);
  }

  private match(...tokenTypes: TokenType[]): boolean {
    for (const tokenType of tokenTypes) {
      if (this.check(tokenType)) {
        this.advance();
        return true;
      }
    }

    return false;
  }

  private consume(tokenType: TokenType, expected: string): Token {
    if (!this.check(tokenType)) {
      throw new SyntaxError(
        this.peek().line,
        this.peek().col,
        `Expected "${expected}", received "${this.peek().lexeme}"`,
      );
    }

    return this.advance();
  }

  private check(tokenType: TokenType): boolean {
    if (this.isAtEnd()) return false;
    return this.peek().token === tokenType;
  }

  private advance(): Token {
    if (!this.isAtEnd()) this.current++;
    return this.previous();
  }

  private isAtEnd() {
    return this.peek().token === TokenType.EOF;
  }

  private peek(): Token {
    return this.tokens[this.current];
  }

  private previous(): Token {
    return this.tokens[this.current - 1];
  }
}

export { SpiceDBSchemaParser };
