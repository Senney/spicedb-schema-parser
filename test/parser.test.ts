import { SpiceDBSchemaLexer } from '../src/lexer';
import { SpiceDBSchemaParser } from '../src/parser';

describe('SpiceDBSchemaParser', () => {
  describe('parse', () => {
    const schema = `
    /** user represents a user */
    definition user {}

    definition group {
      relation member: user
    }

    definition document {
      relation foo: user | group#member
    }
    `;

    test('should match snapshot', () => {
      const tokens = new SpiceDBSchemaLexer(schema).scanAll();

      expect(new SpiceDBSchemaParser(tokens).parse()).toMatchSnapshot();
    });
  });
});
