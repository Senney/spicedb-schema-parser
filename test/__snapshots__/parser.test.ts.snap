// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SpiceDBSchemaParser > parse > should match snapshot 1`] = `
[
  DefinitionStatement {
    "block": BlockStatement {
      "statements": [],
    },
    "resourceName": "user",
  },
  DefinitionStatement {
    "block": BlockStatement {
      "statements": [
        RelationStatement {
          "relationDefinition": RelationDefinitionExpression {
            "expression": IdentifierExpression {
              "relation": Token {
                "col": 109,
                "lexeme": "user",
                "line": 6,
                "literal": "user",
                "token": 0,
              },
            },
          },
          "relationName": "member",
        },
      ],
    },
    "resourceName": "group",
  },
  DefinitionStatement {
    "block": BlockStatement {
      "statements": [
        RelationStatement {
          "relationDefinition": RelationDefinitionExpression {
            "expression": BinaryExpression {
              "lhs": IdentifierExpression {
                "relation": Token {
                  "col": 167,
                  "lexeme": "user",
                  "line": 10,
                  "literal": "user",
                  "token": 0,
                },
              },
              "operator": Token {
                "col": 169,
                "lexeme": "|",
                "line": 10,
                "literal": null,
                "token": 9,
              },
              "rhs": MembershipIdentifierExpression {
                "definition": Token {
                  "col": 175,
                  "lexeme": "group",
                  "line": 10,
                  "literal": "group",
                  "token": 0,
                },
                "relation": Token {
                  "col": 182,
                  "lexeme": "member",
                  "line": 10,
                  "literal": "member",
                  "token": 0,
                },
              },
            },
          },
          "relationName": "foo",
        },
      ],
    },
    "resourceName": "document",
  },
]
`;
